pkg/collector/corechecks/gpu/gpu.go

```diff
@@ -0,0 +1,203 @@
+
+	sysconfig "github.com/DataDog/datadog-agent/cmd/system-probe/config"
+	"github.com/DataDog/datadog-agent/comp/core/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/aggregator/sender"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	pkgconfigsetup "github.com/DataDog/datadog-agent/pkg/config/setup"
+	processnet "github.com/DataDog/datadog-agent/pkg/process/net"
+	sectime "github.com/DataDog/datadog-agent/pkg/security/resolvers/time"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/optional"
+)
+
+// Check doesn't need additional fields
+type Check struct {
+	core.CheckBase
+	config         *CheckConfig
+	sysProbeUtil   *processnet.RemoteSysProbeUtil
+	lastCheckTime  time.Time
+	timeResolver   *sectime.Resolver
+	statProcessors map[uint32]*StatsProcessor
+}
+
+// Factory creates a new check factory
+func Factory() optional.Option[func() check.Check] {
+	return optional.NewOption(newCheck)
+}
+
+func newCheck() check.Check {
+	return &Check{
+		CheckBase:      core.NewCheckBase(CheckName),
+		config:         &CheckConfig{},
+		statProcessors: make(map[uint32]*StatsProcessor),
+	}
+}
+
+// Parse parses the check configuration
+func (m *CheckConfig) Parse(data []byte) error {
+	return yaml.Unmarshal(data, m)
+}
+
+// Cancel cancels the check
+func (m *Check) Cancel() {
+	ret := nvml.Shutdown()
+	if ret != nvml.SUCCESS {
+		log.Warnf("Failed to shutdown NVML: %v", nvml.ErrorString(ret))
+	}
+}
+
+// Configure parses the check configuration and init the check
+func (m *Check) Configure(senderManager sender.SenderManager, _ uint64, config, initConfig integration.Data, source string) error {
+	if err := m.CommonConfigure(senderManager, initConfig, config, source); err != nil {
+		return err
+	}
+	if err := m.config.Parse(config); err != nil {
+		return fmt.Errorf("ebpf check config: %s", err)
+	}
+	if err := processnet.CheckPath(pkgconfigsetup.SystemProbe().GetString("system_probe_config.sysprobe_socket")); err != nil {
+		return fmt.Errorf("sysprobe socket: %s", err)
+	}
+
+	return nil
+}
+
+func (m *Check) ensureInitialized() error {
+	var err error
+
+	if m.sysProbeUtil == nil {
+		m.sysProbeUtil, err = processnet.GetRemoteSystemProbeUtil(
+			pkgconfigsetup.SystemProbe().GetString("system_probe_config.sysprobe_socket"),
+		)
+		if err != nil {
+			return fmt.Errorf("sysprobe connection: %s", err)
+		}
+	}
+
+	if m.timeResolver == nil {
+		m.timeResolver, err = sectime.NewResolver()
+		if err != nil {
+			return fmt.Errorf("cannot create time resolver: %s", err)
+		}
+	}
+	return nil
+}
+
+// ensureProcessor ensures that there is a stats processor for the given key
+func (m *Check) ensureProcessor(key *model.StreamKey, snd sender.Sender, gpuThreads int, checkDuration time.Duration) {
+	if _, ok := m.statProcessors[key.Pid]; !ok {
```

Comment:
Is PID a sufficiently unique key? PIDs can be re-used. I believe if you combine PID with process launch time, then it becomes unique.

Reply:
I think it's sufficiently unique here where we're only monitoring processes that use the GPU. Also, the processor will be deleted from the map if there's no activity from a PID, so the old data wouldn't get reused.